game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def get_last_move(self):
+        return self._moves_list[-1]
+    
+    def get_connect_number(self):
+        return self._connect_number
+
+    def get_number_of_moves(self):
+        return self._number_of_moves
+    
+    def make_move(self, move, player_next = True):
+               
+        self._board[move[0]][move[1]] = self._player_next
+        self._player_next = - self._player_next
+        self._moves_list.append(move)
+        self._number_of_moves = self._number_of_moves + 1
+
+        # return game_state
+        
+    def unmake_move(self):
+        move = self.get_last_move()
+        self._board[move[0]][move[1]] = None
+        self._player_next = - self._player_next

```

Comment:
One of your questions was around using the external-facing getters within your class or not.

I like to opt for consistency first and foremost. Here, I would generally want to use the external getters, but don't mind this way if you're consistent.

The key here is that that class should be small enough that refactoring/changing how it works internally should be quick, but that refactoring it's public-interface should generally be harder and therefore avoided. Of course refactoring the public-interface would still be quite easy for you at this point, but at the other end of a spectrum is if this class was the API into using a library/package. In this case you might be introducing breaking-changes and would need to increment the major-version as appropriate. See here for a quick browse on SemVar: https://semver.org/